src/yarp.c

```diff
@@ -11218,42 +11235,109 @@ parse_expression_prefix(yp_parser_t *parser, yp_binding_power_t binding_power) {
             yp_heredoc_indent_t indent = parser->lex_modes.current->as.heredoc.indent;
 
             yp_node_t *node;
-            if (quote == YP_HEREDOC_QUOTE_BACKTICK) {
-                node = (yp_node_t *) yp_interpolated_xstring_node_create(parser, &parser->current, &parser->current);
-            } else {
-                node = (yp_node_t *) yp_interpolated_string_node_create(parser, &parser->current, NULL, &parser->current);
-            }
 
             parser_lex(parser);
-            yp_node_t *part;
-
-            while (!match_any_type_p(parser, 2, YP_TOKEN_HEREDOC_END, YP_TOKEN_EOF)) {
-                if ((part = parse_string_part(parser)) == NULL) continue;
 
+            if (parser->current.type == YP_TOKEN_HEREDOC_END) {
                 if (quote == YP_HEREDOC_QUOTE_BACKTICK) {
-                    yp_interpolated_xstring_node_append((yp_interpolated_x_string_node_t *) node, part);
+                    node = (yp_node_t *) yp_xstring_node_create(
+                            parser,
+                            &parser->previous,
+                            &((yp_token_t) { .type = YP_TOKEN_STRING_CONTENT, .start = parser->previous.end, .end = parser->current.start }),
+                            &parser->current);
+
                 } else {
-                    yp_interpolated_string_node_append((yp_interpolated_string_node_t *) node, part);
+                    node = (yp_node_t *)yp_string_node_create(
+                            parser,
+                            &parser->previous,
+                            &((yp_token_t) { .type = YP_TOKEN_STRING_CONTENT, .start = parser->previous.end, .end = parser->current.start }),
+                            &parser->current);
                 }
+                node->location = (yp_location_t) {
```

Comment:
From your calls to `yp_string_node_create`, `node->location.start` should already be set correctly. So I think this line can be:

```suggestion
                node->location.end = parser->previous.end;
```